### ISSUE TITLE
State of DataProtocol and FIleProtocol May diverge.

### ISSUE THREAD

**Issue by farhoud** [functionland/fula#202]
Uploading a file in application need to step:
- upload file and get the CID with `fileprotocol`
- record the CID with other data in `graph-protocol` for querying. (like getting list of files)

Scenario ending up with a split ed state:
- If any thing goes wrong with second step we have orphan.
- Later if with add operation like remove we have more splinting state scenarios.
- If a user use IPFS to remove a data we have split-ed state.

**Comment by mehdibalouchi** [comment]
Standardizing the records on the graph protocol makes the API opinionated and it will welcome hacks on the developer side.
Nonetheless, I think there is no way to come up with a standard for storing all the file-related data.
As @farhoud mentioned, the use case here is to add some related records after the file uploading is complete. We have two different types of file-related records here:
- File's metadata: The records that are about the file itself. Right now, the file protocol handles simple forms of metadata like name, modification date, etc.  We can add some additional data to this layer and support directory structures or collections.
- File's relations: These records are for connecting the file (that was just uploaded) to other entities in the application. For example, when you are uploading a photo attachment for a to-do item. These records are tightly coupled with the application's logic. 

We can add support for collections of files as some sort of metadata. This will help the application to create permanent collections and search among them if they want (or get the list of them for the file manager).

In conclusion, I think we cannot (and should not) prevent diverging of the data protocol and the file protocol states. What we can do is provide a way for making permanent collections of files. This should also have the support for nested collections.

**Comment by ehsan6sha** [comment]
@mehdibalouchi But if data protocol state diverges from file protocol state, for example, if someone erases the files but database is still there, or someone erases the database but files are still there, what are the risks?
How can they be synced again?

**Comment by mehdibalouchi** [comment]
@ehsan6sha these issues are also present when you are using a traditional DB with a web2 arch. 

> if someone erases the files but database is still there

If you are storing some file paths (a pointer to a file on a file system) in a database and then for some reason the records get removed from the DB, there is nothing the DB can do about it. A common solution to this problem for centralized databases is to store the file inside the database records (e.g. blob), this is currently possible with existing APIs (with a low performance because orbitDB keeps all the data inside the RAM).

> or someone erases the database but files are still there

In this case, we will have some orphan files (as @farhoud mentioned). By keeping the file and the graph protocols completely independent, we can ensure that any request for a file is made through the IPFS network. Therefore we can rely on the IPFS garbage collector and pinning service to handle the orphan files.


> How can they be synced again?

The application is responsible for keeping the data protocol and the file protocol states synchronized. This is where permanent collections can help. An application should be able to create a permanent collection (with a permanent id) and store files in it. At any time if the application wants to resync the states, it can search through all the files in the collection and update the data protocol state.

**Comment by mehdibalouchi** [comment]
The collections should have a human-readable identifier (e.g. `todo-app-XXX/photos`). The union of `[USER_ID, APP_ID, COLLECTION_ID]` must be unique over the network

**Comment by farhoud** [comment]
@ehsan6sha @keyvan-m-sadeghi @mehdibalouchi 
We can use [UnixFS](https://github.com/ipfs/specs/blob/main/UNIXFS.md) for embedding folder structure of box in ipfs.

**Comment by ehsan6sha** [comment]
@mehdibalouchi the thing is for central web2 model, it is system admins who have access to the data. So divergence will be minimized and will be handled by an expert. Here it is the public who has access and divergence can be a killer to the experience
